File: frontend/src/api/route53.js
import axios from 'axios'

export default {
  getHostZones(data, cb, errorCb) {
    axios.get(`${data.serverPath}/service/route53/hostedzones`)
      .then(function(response){
        cb(response)
      })
      .catch(function(err){
        errorCb(err)
      });
  },
  getRecordSets(data, cb, errorCb) {
    axios.get(`${data.serverPath}/service/route53/recordsets?zone=${data.zone}`)
      .then(function(response){
        cb(response)
      })
      .catch(function(err){
        errorCb(err)
      });
  },
  deleteRecordSet(data, cb, errorCb) {
    axios.delete(`${data.serverPath}/service/route53/recordsets?zone=${data.zone}&recordName=${data.name}&recordType=${data.type}&recordValue=${data.value}&recordTTL=${data.ttl}&alias=${data.alias}&hostzoneID=${data.hostzoneID}`)
      .then(function(response){
        cb(response)
      })
      .catch(function(err){
        errorCb(err)
      });
  },
  createRecordSet(data, cb, errorCb) {
    axios.post(`${data.serverPath}/service/route53/recordsets?zone=${data.zone}&recordName=${data.name}&recordType=${data.type}&recordValue=${data.value}&recordTTL=${data.ttl}&alias=${data.alias}&hostzoneID=${data.hostzoneID}`)
      .then(function(response){
        cb(response)
      })
      .catch(function(err){
        errorCb(err)
      });
  },
}

File: backend/services/route53.go
package services

import (
	"github.com/aws/aws-sdk-go/service/route53"
)

func GetHostedZones(route53Svc *route53.Route53) (result *route53.ListHostedZonesOutput, error error) {
	input := &route53.ListHostedZonesInput{}
	data, err := route53Svc.ListHostedZones(input)
	if err != nil {
		return nil, err
	}
	return data, nil
}

func GetRecordSets(route53Svc *route53.Route53, zone string) (result *route53.ListResourceRecordSetsOutput, error error) {
	input := &route53.ListResourceRecordSetsInput{
		HostedZoneId: &zone,
	}
	data, err := route53Svc.ListResourceRecordSets(input)
	if err != nil {
		return nil, err
	}
	truncated := *data.IsTruncated
	nextRecordName := data.NextRecordName
	for truncated {
		result, err = route53Svc.ListResourceRecordSets(&route53.ListResourceRecordSetsInput{
			HostedZoneId:    &zone,
			StartRecordName: nextRecordName,
		})
		if err != nil {
			return nil, err
		}
		data.ResourceRecordSets = append(data.ResourceRecordSets, result.ResourceRecordSets...)
		truncated = *result.IsTruncated
		nextRecordName = result.NextRecordName
	}
	return data, nil
}

func ChangeRecordSet(
	route53Svc *route53.Route53, zone string,
	recordName string, recordType string,
	recordValue string, ttl int64,
	action string, alias bool, recordHostZoneID string) (result *route53.ChangeResourceRecordSetsOutput, error error) {
	input := &route53.ChangeResourceRecordSetsInput{
		HostedZoneId: &zone,
		ChangeBatch: &route53.ChangeBatch{
			Changes: []*route53.Change{
				{
					Action: &action,
				},
			},
		},
	}
	if !alias {
		input.ChangeBatch.Changes[0].ResourceRecordSet = &route53.ResourceRecordSet{
			Name: &recordName,
			Type: &recordType,
			ResourceRecords: []*route53.ResourceRecord{
				{
					Value: &recordValue,
				},
			},
			TTL: &ttl,
		}
	} else {
		evaluateTargetHealth := false
		input.ChangeBatch.Changes[0].ResourceRecordSet = &route53.ResourceRecordSet{
			Name: &recordName,
			Type: &recordType,
			AliasTarget: &route53.AliasTarget{
				DNSName:              &recordValue,
				HostedZoneId:         &recordHostZoneID,
				EvaluateTargetHealth: &evaluateTargetHealth,
			},
		}
	}
	data, err := route53Svc.ChangeResourceRecordSets(input)
	if err != nil {
		return nil, err
	}
	return data, nil
}
